### PR TITLE
Provide more context on recursive `impl` evaluation overflow

### LINKED ISSUE #116925: Confusing error when referencing associated types inside trait bounds in trait implementations
### Code

```Rust
trait A<T> {}

trait B {
    type Type;
}

impl<T> B for T
where
    T: A<Self::Type>,
{
    type Type = bool;
}
```

### Current output

```Shell
error[E0275]: overflow evaluating the requirement `<T as B>::Type == <T as B>::Type`
 --> src/lib.rs:7:1
  |
7 | / impl<T> B for T
8 | | where
9 | |     T: A<Self::Type>,
  | |_____________________^
  |
note: required for `T` to implement `B`
 --> src/lib.rs:7:9
  |
7 | impl<T> B for T
  |         ^     ^
8 | where
9 |     T: A<Self::Type>,
  |        ------------- unsatisfied trait bound introduced here

For more information about this error, try `rustc --explain E0275`.
```

### Desired output

```Shell
an explanation that `Self::Type` does not reference `bool` but tries to figure out whether T already implements B
```

### Rationale and extra context

spawned from #116914

### Other cases

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by estebank**
When an associated type `Self::Assoc` is part of a `where` clause, we end up unable to evaluate the requirement and emit a E0275.

We now point at the associated type if specified in the `impl`. If so, we also suggest using that type instead of `Self::Assoc`. Otherwise, we explain that these are not allowed.

```
error[E0275]: overflow evaluating the requirement `<(T,) as Grault>::A == _`
  --> $DIR/impl-wf-cycle-1.rs:15:1
   |
LL | / impl<T: Grault> Grault for (T,)
LL | |
LL | | where
LL | |     Self::A: Baz,
LL | |     Self::B: Fiz,
   | |_________________^
LL |   {
LL |       type A = ();
   |       ------ associated type `<(T,) as Grault>::A` is specified here
   |
note: required for `(T,)` to implement `Grault`
  --> $DIR/impl-wf-cycle-1.rs:15:17
   |
LL | impl<T: Grault> Grault for (T,)
   |                 ^^^^^^     ^^^^
...
LL |     Self::A: Baz,
   |              --- unsatisfied trait bound introduced here
   = note: 1 redundant requirement hidden
   = note: required for `(T,)` to implement `Grault`
help: associated type for the current `impl` cannot be restricted in `where` clauses, remove this bound
   |
LL -     Self::A: Baz,
   |
```
```
error[E0275]: overflow evaluating the requirement `<T as B>::Type == <T as B>::Type`
  --> $DIR/impl-wf-cycle-3.rs:7:1
   |
LL | / impl<T> B for T
LL | | where
LL | |     T: A<Self::Type>,
   | |_____________________^
LL |   {
LL |       type Type = bool;
   |       --------- associated type `<T as B>::Type` is specified here
   |
note: required for `T` to implement `B`
  --> $DIR/impl-wf-cycle-3.rs:7:9
   |
LL | impl<T> B for T
   |         ^     ^
LL | where
LL |     T: A<Self::Type>,
   |        ------------- unsatisfied trait bound introduced here
help: replace the associated type with the type specified in this `impl`
   |
LL |     T: A<bool>,
   |          ~~~~
```
```
error[E0275]: overflow evaluating the requirement `<T as Filter>::ToMatch == <T as Filter>::ToMatch`
  --> $DIR/impl-wf-cycle-4.rs:5:1
   |
LL | / impl<T> Filter for T
LL | | where
LL | |     T: Fn(Self::ToMatch),
   | |_________________________^
   |
note: required for `T` to implement `Filter`
  --> $DIR/impl-wf-cycle-4.rs:5:9
   |
LL | impl<T> Filter for T
   |         ^^^^^^     ^
LL | where
LL |     T: Fn(Self::ToMatch),
   |        ----------------- unsatisfied trait bound introduced here
note: associated types for the current `impl` cannot be restricted in `where` clauses
  --> $DIR/impl-wf-cycle-4.rs:7:11
   |
LL |     T: Fn(Self::ToMatch),
   |           ^^^^^^^^^^^^^
```

Fix #116925

### CODE REVIEW

**Comment by petrochenkov** [comment]
r? compiler
I'm overloaded with reviews.

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 1]
```diff
@@ -5052,6 +5074,136 @@ pub fn suggest_desugaring_async_fn_to_impl_future_in_trait<'tcx>(
...
+                        .predicates
+                        .iter()
+                        .filter(|p| {
+                            if let hir::WherePredicate::BoundPredicate(p) = p
+                                && hir::PredicateOrigin::WhereClause == p.origin
+                            {
+                                true
+                            } else {
+                                false
+                            }
```
Comment (TaKO8Ki) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:5129]: nit: 
```suggestion
                            matches!(hir::WherePredicate::BoundPredicate(p), p if hir::PredicateOrigin::WhereClause == p.origin)
```

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 2]
```diff
@@ -5052,6 +5074,136 @@ pub fn suggest_desugaring_async_fn_to_impl_future_in_trait<'tcx>(
...
+                );
+                // Search for the associated type `Self::{name}`, get
+                // its type and suggest replacing the bound with it.
+                let mut visitor = SelfVisitor { paths: vec![], name: Some(name) };
+                visitor.visit_trait_ref(trait_ref);
+                for path in visitor.paths {
+                    err.span_suggestion_verbose(
+                        path.span,
+                        "replace the associated type with the type specified in this `impl`",
+                        format!("{}", tcx.type_of(new.def_id).skip_binder(),),
```
Comment (TaKO8Ki) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:5187]: nit:

```suggestion
                        tcx.type_of(new.def_id).skip_binder().to_string()
```

**Review by TaKO8Ki — COMMENTED** [review]
I left some suggestions. Other than them, r=me

**Pull request merged**